controllers/external/util.go

```diff
@@ -88,30 +90,143 @@ type CreateFromTemplateInput struct {
+
+	// If UID is specified it will be used on the target object.
+	// +optional
+	UID types.UID
+
+	// Namespace is the Kubernetes namespace the cloned object should be created into.
+	Namespace string
+
+	// ClusterName is the cluster this object is linked to.
+	ClusterName string
+
+	// OwnerRef is an optional OwnerReference to attach to the cloned object.
+	// +optional
+	OwnerRef *metav1.OwnerReference
+
+	// Labels is an optional map of labels to be added to the object.
+	// +optional
+	Labels map[string]string
+
+	// Annotations is an optional map of annotations to be added to the object.
+	// +optional
+	Annotations map[string]string
+
+	// Manager is the name of the manager to when using SSA patch.
+	Manager string
+}
+
+// ApplyFromTemplate uses the client and the reference to create a new object from the template.
+func ApplyFromTemplate(ctx context.Context, in *ApplyFromTemplateInput) (*corev1.ObjectReference, error) {
+	if in.Manager == "" {
+		return nil, fmt.Errorf("manager is not specified")
+	}
+	obj, err := ComputeObjectFromTemplate(ctx, &ComputeObjectFromTemplateInput{
+		Client:      in.Client,
 		TemplateRef: in.TemplateRef,
 		Namespace:   in.Namespace,
 		ClusterName: in.ClusterName,
 		OwnerRef:    in.OwnerRef,
 		Labels:      in.Labels,
 		Annotations: in.Annotations,
-	}
-	to, err := GenerateTemplate(generateTemplateInput)
+	})
 	if err != nil {
+		return nil, errors.Wrap(err, "failed to compute object form template")
+	}
+	if in.Name != "" {
+		obj.SetName(in.Name)
+	}
+	obj.SetUID(in.UID)
+
+	// Apply the patch using SSA.
+	patchOptions := []client.PatchOption{
+		client.ForceOwnership,
+		client.FieldOwner(in.Manager),
+	}
+	if err := in.Client.Patch(ctx, obj, client.Apply, patchOptions...); err != nil {
 		return nil, err
 	}
 
-	// Create the external clone.
-	if err := in.Client.Create(ctx, to); err != nil {
+	return GetObjectReference(obj), nil
+}
+
+// ComputeObjectFromTemplateInput is the input to ComputeObjectFromTemplate.
+type ComputeObjectFromTemplateInput struct {
+	// Client is the controller runtime client.
+	Client client.Client
+
+	// TemplateRef is a reference to the template that needs to be cloned.
+	TemplateRef *corev1.ObjectReference
+
+	// Namespace is the Kubernetes namespace the cloned object should be created into.
+	Namespace string
+
+	// ClusterName is the cluster this object is linked to.
+	ClusterName string
+
+	// OwnerRef is an optional OwnerReference to attach to the cloned object.
+	// +optional
+	OwnerRef *metav1.OwnerReference
+
+	// Labels is an optional map of labels to be added to the object.
+	// +optional
+	Labels map[string]string
+
+	// Annotations is an optional map of annotations to be added to the object.
+	// +optional
+	Annotations map[string]string
+}
+
+// ComputeObjectFromTemplate computes the target object from the provided template.
+func ComputeObjectFromTemplate(ctx context.Context, in *ComputeObjectFromTemplateInput) (*unstructured.Unstructured, error) {
```

Comment:
If I got this right we are introducing this func in order to not call two times Get, not sure it is worth the effort.
I will keep things simple and avoid introducing this layer on top of GenerateTemplate

Reply:
+1